app.js

```diff
@@ -103,12 +166,20 @@ app.get("/disconnect", (req, res) => {
 
 app.ws("/", ws => {
   ws.on("message", () => {
-    ws.send(JSON.stringify(settings.onEvent.message));
+    settingPersister.copy((err, copiedData) => {
+      // TODO: error handling?
+      if (javaScriptUtils.isDefined(copiedData.onEventMessage)) {
+        ws.send(copiedData.onEventMessage);
+      }
+    });
   });
 
-  if (settings.isPeriodicMessageSendingActive) {
-    createPeriodicMessageInterval(ws);
-  }
+  settingPersister.copy((err, copiedData) => {
+    // TODO: error handling?
```

Comment:
We don't need to handle errors unless there's a strategy on how to handle them. Also, we're not building critical software. With that in mind consider the following points when handling errors:
- **Catching errors and letting the app continue often will result in an error / inconsistent state later on which might be even harded to debug. Let the app crash unless you know how to continue the app logic with a certain operation failed.**
- Catching the error just to log them instead of letting them fail doesn't make sense since exceptions will be logged automatically.
- While promises provide a `catch` method, most of them will never fail unless there's a bug. If there's a bug we don't need to handle it gracefully, let the app crash, so we notice and can fix it. There're things which are nice to handle and can fail more often (e.g. network calls) but also in that case you need a strategy to handle them (e.g. retry). Accessing the DB imo we can consider safe and we shouldn't handle errors.

My 2 cents from a different background, but if you could read up on this topic regarding JS and promises that'd be welcome.

Reply:
Wow, pretty concise description, thanks. :) All of the listed points are reasonable.

That's a myth for me how people handle errors gracefully - e.g. what the best practices or use cases are. I'll try to digest deeper into the details as it's both quite interesting and also should be useful as a future reference.

That being said, let's remove all the error handling related lines until we lay down a common consensus. Furthermore, the details can also be documented as a convention in our case.

Reply:
First off, let's restrict the discussion to sync and callback-based code. It is what's relevant here and should be easier to grasp. I've edited a point in my answer above. I will also add that **fail early** is a general best practice.

Couldn't find anything that addresses all aspects of this, but [this article](https://medium.com/datafire-io/es6-promises-patterns-and-anti-patterns-bbb21a5d0918) makes some good points about "Not catching errors" (as an anti-pattern). `if (err) throw err;` seems like a reasonable approach for libraries that use "errbacks", like `mongoose`. Let's go with that.

Some actual comments:
> `if (err) return console.error(err);`

Not a good idea, because the caller might continue assuming no error. These many times result in some `undefined` error somewhere down the line. Fail early. :)

> if (err) throw new Error("Incorrect MongoDB connection URI - " + err);

Not sure what converting `err` to string does, but it many cases it can hide information. And this here will definitely hide the original trace. Also, you probably shouldn't assume the exact error unless you can [check it](https://stackoverflow.com/a/30469297/2771889). So (if possible) you could do this:
```
if (err instanceof MongooseInvalidUriError)
  throw new Error("Incorrect MongoDB connection URI);
else if (err)
  throw err;
end
```

> `if (err) { return res.status(200).json({ success: false }) }`

... :D Also, an exception in a sync case usually already takes care of returning an error response, so one more reason for failing hard. `{ success: false }` is redundant if there's a proper response code.

Reply:
The `err.name` property holds some useful logic of a particular error. Converting the example pseudo code to JS would be similar to the one below:

```js
if (err.name === "MongoNetworkError") {
   // ...
}
```

Another approach might be to create [**custom error objects**](https://medium.com/@xjamundx/custom-javascript-errors-in-es6-aa891b173f87) - not sure if it's feasible in our case.

Reply:
As we couldn't decide what'd be the reasonable approach to handle async errors, the related `catch` blocks should be deleted, shouldn't it?

Reply:
> My concern is that it's pretty static though - e.g. the compared string can be mistyped.

That'd be a bug, but if written correctly is should work correctly.

> Another approach might be to create custom error objects

When we're throwing errors, that's a good approach. The bigger concern I think is catching errors.

> As we couldn't decide what'd be the reasonable approach to handle async errors, the related catch blocks should be deleted, shouldn't they?

No, your catch blocks are all well handled from what I can see.